File: Data Structures & Algorithms/binary-search/submission-4.java
class Solution {
    public int search(int[] nums, int target) {
        
        int l = 0;
        int r = nums.length - 1;
        // 1 2 3 4 6
        // l   m   r
        //     l m r
        //       l r  
        while (l <= r) {
            int mid = (l + r) / 2;
            int current = nums[mid];

            if (current == target) {
                return mid;
            }

            if (current < target) {
                l = mid + 1;
            }

            if (current > target) {
                r = mid - 1;
            }

        }

        return -1;
    }
}

File: Data Structures & Algorithms/permutations/submission-4.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        res = []
        used = [False] * len(nums)

        def backtrack(path):
            # Base Case: All slots are filled
            if len(path) == len(nums):
                res.append(path.copy()) # Save a copy of the completed permutation
                return

            # Recursive Case: Try every available number for the current slot
            for i in range(len(nums)):
                if not used[i]:
                    # 1. Make a decision
                    used[i] = True
                    path.append(nums[i])

                    # 2. Explore deeper (fill the next slot)
                    backtrack(path)

                    # 3. Undo the decision (Backtrack)
                    path.pop()
                    used[i] = False

        backtrack([])
        return res